### PR TITLE
Parachain inherents rebase

### PR DESCRIPTION
**Pull request by cdamian**
Re-base for the work done in - https://github.com/centrifuge/fudge/pull/28

### CODE REVIEW

File: core/src/builder/parachain.rs [hunk 1]
```diff
@@ -57,6 +106,251 @@ pub struct FudgeParaChain {
...
+				})?
+		};
+
+		Ok(Some(collation_info))
+	}
+
+	pub fn collation(&self, validation_data: PersistedValidationData) -> Option<Collation> {
+		let _state = self.backend.state_at(self.client.info().best_hash).ok()?;
+		//ExternalitiesProvider::<HashFor<Block>, B::State>::new(&state)
+		//	.execute_with(|| self.create_collation(validation_data))
```
Comment (mustermeiszer) [inline core/src/builder/parachain.rs:222]: I think we can remove that.

File: core/src/builder/parachain.rs [hunk 2]
```diff
@@ -185,103 +511,298 @@ where
...
+			tracing::error!(
+				target = DEFAULT_PARACHAIN_BUILDER_LOG_TARGET,
+				error = ?e,
+				"Lock for next block is poisoned.",
+			);
+
+			Error::NextBlockLockPoisoned(InnerError::from(e.to_string()))
+		})?;
+
+		*next_block = Some((block, proof));
```
Comment (mustermeiszer) [inline core/src/builder/parachain.rs:606]: We might need to consider racing conditions between the collator and the builder if we make the whole stuff async. But for now it should work fine.

File: core/src/builder/parachain.rs [hunk 3]
```diff
@@ -185,103 +511,298 @@ where
...
+			tracing::error!(
+				target = DEFAULT_PARACHAIN_BUILDER_LOG_TARGET,
+				error = ?e,
+				"Lock for next import is poisoned.",
+			);
+
+			Error::NextImportLockPoisoned(InnerError::from(e.to_string()))
+		})?;
+
+		*next_import = block;
```
Comment (mustermeiszer) [inline core/src/builder/parachain.rs:639]: I think we can remove that one here.

**Review by mustermeiszer — COMMENTED** [review]
Just some minor stuff. Otherwise looks sick! Thanks for taking care of this. Was quite some work ^^

Couldn't find the code where we alter the relay-chain storage to actually not panic if there are not authoritites. Could you guide me ther?

File: core/src/builder/parachain.rs (hunk 1 above)
Comment (cdamian) [reply core/src/builder/parachain.rs:222]: Done.

File: core/src/builder/parachain.rs (hunk 2 above)
Comment (cdamian) [reply core/src/builder/parachain.rs:606]: The collator also holds a `Arc<Mutex<Option<(Block, StorageProof)>>>`, I think we are safe in that regard,

**Comment by cdamian** [comment]
> Just some minor stuff. Otherwise looks sick! Thanks for taking care of this. Was quite some work ^^
> 
> Couldn't find the code where we alter the relay-chain storage to actually not panic if there are not authoritites. Could you guide me ther?

I assume you are referring to [this](https://github.com/centrifuge/fudge/blob/parachain-inherents-rebase/core/src/builder/relay_chain.rs#L628)? ^^

File: core/src/builder/parachain.rs (hunk 2 above)
Comment (mustermeiszer) [reply core/src/builder/parachain.rs]: From a computational point that is true. I am just not sure, if the internal logic has problems with that.

**Comment by mustermeiszer** [comment]
> I assume you are referring to [this](https://github.com/centrifuge/fudge/blob/parachain-inherents-rebase/core/src/builder/relay_chain.rs#L628)? ^^

Yes, thanks. ^^ 

This should be rather in the genesis startup and not for each parachain. Also can leave a todo that we need to have a way to actually renew that infro for each session or have some test to carry it over multiple sessions?

**Comment by cdamian** [comment]
> > I assume you are referring to [this](https://github.com/centrifuge/fudge/blob/parachain-inherents-rebase/core/src/builder/relay_chain.rs#L628)? ^^
> 
> Yes, thanks. ^^
> 
> This should be rather in the genesis startup and not for each parachain. Also can leave a todo that we need to have a way to actually renew that infro for each session or have some test to carry it over multiple sessions?

How would we add that during genesis? The `session_info` pallet does not have a `GenesisConfig` and the `AccountKeys` storage is not visible outside of that crate.

From what I see in the `session_info` pallet, `initializer_on_new_session` should populate that storage, when called from the `initializer` pallet. But unfortunately I was not able to debug this so I can get more knowledge on this whole flow.

**Comment by mustermeiszer** [comment]
Good argument TBH. Haha. Well okay, then let's see what my research regarding the actual flow brings and otherwise, keep it as is.

**Comment by cdamian** [comment]
> Good argument TBH. Haha. Well okay, then let's see what my research regarding the actual flow brings and otherwise, keep it as is.

Played around with `schedule_para_initialize` to see if that would work, since it would trigger all the needed logic from what I see, but it seems that we can't do pallet calls from the context of the relaychain builder it seems. I get an error when the pallet tries to read something from storage...

**Pull request merged**